app/main/views/suppliers.py

```diff
@@ -512,25 +513,49 @@ def create_new_supplier():
 def duns_number():
     form = DunsNumberForm()
 
-    if request.method == 'POST':
-        if form.validate_on_submit():
-            suppliers = data_api_client.find_suppliers(duns_number=form.duns_number.data)
-            if len(suppliers["suppliers"]) <= 0:
-                session[form.duns_number.name] = form.duns_number.data
-                return redirect(url_for(".company_details"))
+    if request.method == 'GET':
+        if request.args.get('retry', None) == 'true':
+            # When redirected back from the lookup page
+            form.duns_number.errors = ["Enter a different DUNS number"]
+        elif form.duns_number.name in session:
+            # Likely a back click or unexpected navigation, refill form data
+            form.duns_number.data = session[form.duns_number.name]
+
+    if form.validate_on_submit():
+        # Valid duns number
 
+        # Fail on existing account with this duns number first
+        if len(data_api_client.find_suppliers(duns_number=form.duns_number.data)['suppliers']) > 0:
             form.duns_number.errors = ["DUNS number already used"]
+        else:
+            # Check for number in Direct Plus API
+            # If we cannot contact the API then skip this step
+            # If the client doesn't return an organization then there is not a record for that DUNs
+            # Otherwise the number is good, add the 'primaryName' to the session and continue
+            try:
+                organization = direct_plus_client.get_organization_by_duns_number(form.duns_number.data)
+                if organization is not None:
+                    company_name = organization['primaryName']
+            except KeyError:
+                # An unexpected error. Something other than supplier data in the response. Skip this part of sign up.
+                return redirect(url_for(".company_details"))
+
+            organization = direct_plus_client.get_organization_by_duns_number(form.duns_number.data)
```

Comment:
Why are we doing this call again when we've potentially just fallen out of the above `try` block where we already do this and why is this call not in a `try` block? What happens if this one fails?

Reply:
Good catch, merge conflict thing I think